File: apps/api/src/lib/dedupe.ts
// Deterministic dedupe key helper — sha256(sourceId + normalized title).
// Used for idempotent re-extraction in Phase 2.

import { createHash } from "crypto";

export function dedupeKey(sourceId: string, title: string): string {
  const norm = title.trim().toLowerCase().replace(/\s+/g, " ");
  return createHash("sha256").update(`${sourceId}|${norm}`).digest("hex");
}

File: apps/api-py/app/core/dedupe.py
"""Deterministic dedupe key — sha256(sourceId + normalized title).

Port of ``apps/api/src/lib/dedupe.ts``. Used for idempotent re-extraction.
"""
from __future__ import annotations

import hashlib
import re

_WS = re.compile(r"\s+")


def dedupe_key(source_id: str, title: str) -> str:
    norm = _WS.sub(" ", title.strip().lower())
    return hashlib.sha256(f"{source_id}|{norm}".encode("utf-8")).hexdigest()
